pkg/runtime/resolver.go

```diff
@@ -0,0 +1,214 @@
+
+	// defaultConfigNamespace defines the namespace that contains configuration defaults for istio.
+	// This is distinct from the "default" namespace in K8s.
+	// default: istio-default-config
+	defaultConfigNamespace string
+
+	// rules in the configuration database keyed by $namespace.
+	rules map[string][]*Rule
+
+	// Opaque ruleState used in this configuration
+	// After the resolver has no references, this state can be
+	// cleaned up correctly by the caller.
+	ruleState interface{}
+
+	// refCount tracks the number requests currently using this
+	// configuration. resolver state can be cleaned up when this count is 0.
+	refCount int32
+}
+
+// NewResolver returns a Resolver.
+func NewResolver(evaluator expr.PredicateEvaluator, identityAttribute string, defaultConfigNamespace string,
+	rules map[string][]*Rule, ruleState interface{}) Resolver {
+	return &resolver{
+		evaluator:              evaluator,
+		identityAttribute:      identityAttribute,
+		defaultConfigNamespace: defaultConfigNamespace,
+		rules:     rules,
+		ruleState: ruleState,
+	}
+}
+
+// DefaultConfigNamespace holds istio wide configuration.
+const DefaultConfigNamespace = "istio-default-config"
+
+// DefaultIdentityAttribute is attribute that defines config scopes.
+const DefaultIdentityAttribute = "target.service"
+
+// expectedResolvedActionsCount is used to preallocate slice for actions.
+const expectedResolvedActionsCount = 10
+
+// Resolve resolves the in memory configuration to a set of actions based on request attributes.
+// Resolution is performed in the following order
+// 1. Check rules from the defaultConfigNamespace -- these rules always apply
+// 2. Check rules from the target.service namespace
+func (r *resolver) Resolve(attrs attribute.Bag, variety adptTmpl.TemplateVariety) (ra Actions, err error) {
+	nselected := 0
+	target := "unknown"
+
+	start := time.Now()
+	// increase refcount just before returning
+	// only if there is no error.
+	defer func() {
+		if err == nil {
+			r.incRefCount()
+		}
+	}()
+
+	// monitoring info
+	defer func() {
+		lbls := prometheus.Labels{
+			targetStr: target,
+			errorStr:  strconv.FormatBool(err != nil),
+		}
+		resolveCounter.With(lbls).Inc()
+		resolveDuration.With(lbls).Observe(time.Since(start).Seconds())
+		resolveRules.With(lbls).Observe(float64(nselected))
+		raLen := 0
+		if ra != nil {
+			raLen = len(ra.Get())
+		}
+		resolveActions.With(lbls).Observe(float64(raLen))
+	}()
+
+	//
+	attr, _ := attrs.Get(r.identityAttribute)
+	if attr == nil {
+		msg := fmt.Sprintf("%s identity not found in attributes%v", r.identityAttribute, attrs.Names())
+		glog.Warningf(msg)
+		return nil, errors.New(msg)
+	}
+
+	rulesArr := make([][]*Rule, 0, 2)
+
+	// add default namespace if present
+	if dcf := r.rules[r.defaultConfigNamespace]; dcf != nil {
+		rulesArr = append(rulesArr, dcf)
+	} else if glog.V(3) {
+		glog.Infof("Resolve: no namespace config for %s", r.defaultConfigNamespace)
+	}
+
+	target = attr.(string)
```

Comment:
good point. we should not check it here though. Identity attribute is part of the configuration, and we should fail validation if that attribute is ...`request.size`.

Practically that attribute is `destination.service` or (`target.service`) in the old nomenclature.